In [2]:
import pandas as pd

Wir lesen hier unsere CSV-Datei ein, die die Daten der letzten Bundesliga Saisons (ab 19/20) hat

In [3]:
matches = pd.read_csv('../datasets/merge.csv')

Mit matches.head() lassen wir uns die ersten 5 Zeilen der Tabelle anzeigen

In [4]:
matches.head()

Div        Date   Time       HomeTeam            AwayTeam  FTHG  FTAG FTR  \
0  D1  16/08/2019  19:30  Bayern Munich              Hertha     2     2   D   
1  D1  17/08/2019  14:30       Dortmund            Augsburg     5     1   H   
2  D1  17/08/2019  14:30       Freiburg               Mainz     3     0   H   
3  D1  17/08/2019  14:30     Leverkusen           Paderborn     3     2   H   
4  D1  17/08/2019  14:30  Werder Bremen  Fortuna Dusseldorf     1     3   A   

   HTHG  HTAG  ... AvgC<2.5  AHCh  B365CAHH  B365CAHA  PCAHH  PCAHA  MaxCAHH  \
0     1     2  ...     3.40 -2.25      2.03      1.90   1.99   1.93     2.04   
1     1     1  ...     3.31 -2.25      1.92      2.01   1.92   2.00     1.98   
2     0     0  ...     1.89  0.00      1.92      2.01   1.94   1.97     1.97   
3     2     2  ...     3.58 -2.00      2.07      1.86   2.05   1.86     2.15   
4     0     1  ...     2.26 -0.75      1.92      2.01   1.92   2.00     1.95   

   MaxCAHA  AvgCAHH  AvgCAHA  
0     1.93     1.98     1.91  
1     2.04     1.91     1.97  
2     2.06     1.90     1.99  
3     1.91     2.03     1.85  
4     2.11     1.89        2  

[5 rows x 105 columns]

matches.shape zeigt uns die Länge der x- und y-Achse. Wir haben 105 Features (Spalten) die wir für unsere Prediction theoretisch verwenden können.

In [5]:
matches.shape

(1332, 105)

Wir schauen uns an wie häufig die Heimmanschaften in der Tabelle auftauchen.

In [6]:
matches["HomeTeam"].value_counts()

HomeTeam
Dortmund              75
Bayern Munich         74
Ein Frankfurt         74
RB Leipzig            74
Augsburg              74
FC Koln               74
Union Berlin          74
Hoffenheim            74
M'gladbach            74
Wolfsburg             74
Leverkusen            74
Freiburg              74
Mainz                 73
Hertha                68
Werder Bremen         58
Stuttgart             57
Schalke 04            51
Bochum                39
Bielefeld             34
Fortuna Dusseldorf    17
Paderborn             17
Greuther Furth        17
Darmstadt              6
Heidenheim             6
Name: count, dtype: int64

Wir lassen uns die Heimspiele von Bayern anzeigen

In [7]:
matches[matches["HomeTeam"] == "Bayern Munich"]

Div        Date   Time       HomeTeam      AwayTeam  FTHG  FTAG FTR  \
0     D1  16/08/2019  19:30  Bayern Munich        Hertha     2     2   D   
19    D1  31/08/2019  14:30  Bayern Munich         Mainz     6     1   H   
37    D1  21/09/2019  14:30  Bayern Munich       FC Koln     4     0   H   
55    D1  05/10/2019  14:30  Bayern Munich    Hoffenheim     1     2   A   
73    D1  26/10/2019  14:30  Bayern Munich  Union Berlin     2     1   H   
...   ..         ...    ...            ...           ...   ...   ...  ..   
1251  D1  15/09/2023  19:30  Bayern Munich    Leverkusen     2     2   D   
1262  D1  23/09/2023  14:30  Bayern Munich        Bochum     7     0   H   
1285  D1  08/10/2023  16:30  Bayern Munich      Freiburg     3     0   H   
1298  D1  28/10/2023  14:30  Bayern Munich     Darmstadt     8     0   H   
1316  D1  11/11/2023  14:30  Bayern Munich    Heidenheim     4     2   H   

      HTHG  HTAG  ... AvgC<2.5  AHCh  B365CAHH  B365CAHA  PCAHH  PCAHA  \
0        1     2  ...     3.40 -2.25      2.03      1.90   1.99   1.93   
19       2     1  ...     4.32 -2.75      2.05      1.88   2.03   1.88   
37       1     0  ...     4.07 -2.50      1.86      2.07   1.86   2.06   
55       0     0  ...     4.91 -2.75      1.94      1.99   1.92   2.00   
73       1     0  ...     4.49 -2.75      1.96      1.97   1.95   1.96   
...    ...   ...  ...      ...   ...       ...       ...    ...    ...   
1251     1     1  ...     3.09 -1.00      1.95      1.98   1.96   1.96   
1262     4     0  ...     4.38 -2.50      2.01      1.92   2.01   1.90   
1285     2     0  ...     4.18 -2.50      1.75      2.05   1.77   2.13   
1298     0     0  ...     5.31 -3.00      1.90      2.03   1.85   2.05   
1316     2     0  ...     4.73 -2.75      2.03      1.90   2.00   1.89   

      MaxCAHH  MaxCAHA  AvgCAHH  AvgCAHA  
0        2.04     1.93     1.98     1.91  
19       2.11     1.90     2.02     1.86  
37       1.90     2.08     1.86     2.02  
55       1.98     2.03     1.93     1.95  
73       1.99     1.99     1.95     1.93  
...       ...      ...      ...      ...  
1251     2.00     2.04     1.92     1.94  
1262     2.04     2.02     1.94      1.9  
1285     2.10     2.19     1.86        2  
1298     1.96     2.05     1.89     1.96  
1316     2.04     2.14     1.92     1.92  

[74 rows x 105 columns]

Mit .dtypes können wir uns die Datentypen der einzelnen Spalten anzeigen lassen. Das ist wichtig, weil manche Parameter als Objects gespeichert sind, sie wir aber als Zahl benötigen, da ein ML-Model wenig mit Objekten anfangen kann.

In [8]:
matches.dtypes

Div          object
Date         object
Time         object
HomeTeam     object
AwayTeam     object
             ...   
PCAHA       float64
MaxCAHH     float64
MaxCAHA     float64
AvgCAHH     float64
AvgCAHA      object
Length: 105, dtype: object

Die Spalte Div ist unnötig, weil sie nur anzeigt um welche Liga es geht. Mit dem Drop dieser Spalte sparen wir uns Zeit beim iterieren der Datei.

In [9]:
matches.drop(["Div"], axis="columns", inplace=True)

In [10]:
matches.head()

Date   Time       HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
0  16/08/2019  19:30  Bayern Munich              Hertha     2     2   D     1   
1  17/08/2019  14:30       Dortmund            Augsburg     5     1   H     1   
2  17/08/2019  14:30       Freiburg               Mainz     3     0   H     0   
3  17/08/2019  14:30     Leverkusen           Paderborn     3     2   H     2   
4  17/08/2019  14:30  Werder Bremen  Fortuna Dusseldorf     1     3   A     0   

   HTAG HTR  ...  AvgC<2.5  AHCh  B365CAHH  B365CAHA  PCAHH  PCAHA  MaxCAHH  \
0     2   A  ...      3.40 -2.25      2.03      1.90   1.99   1.93     2.04   
1     1   D  ...      3.31 -2.25      1.92      2.01   1.92   2.00     1.98   
2     0   D  ...      1.89  0.00      1.92      2.01   1.94   1.97     1.97   
3     2   D  ...      3.58 -2.00      2.07      1.86   2.05   1.86     2.15   
4     1   A  ...      2.26 -0.75      1.92      2.01   1.92   2.00     1.95   

   MaxCAHA  AvgCAHH  AvgCAHA  
0     1.93     1.98     1.91  
1     2.04     1.91     1.97  
2     2.06     1.90     1.99  
3     1.91     2.03     1.85  
4     2.11     1.89        2  

[5 rows x 104 columns]

Hier wandeln wir die Einträge in der Spalte "Datum" in ein passendes Format um, das vom Model gelesen werden kann.

In [11]:
matches["Date"] = pd.to_datetime(matches["Date"])

/var/folders/lw/pygk9wrs0711t4dprf56xt_m0000gn/T/ipykernel_7708/254043194.py:1: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  matches["Date"] = pd.to_datetime(matches["Date"])


In [12]:
matches.head()

Date   Time       HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
0 2019-08-16  19:30  Bayern Munich              Hertha     2     2   D     1   
1 2019-08-17  14:30       Dortmund            Augsburg     5     1   H     1   
2 2019-08-17  14:30       Freiburg               Mainz     3     0   H     0   
3 2019-08-17  14:30     Leverkusen           Paderborn     3     2   H     2   
4 2019-08-17  14:30  Werder Bremen  Fortuna Dusseldorf     1     3   A     0   

   HTAG HTR  ...  AvgC<2.5  AHCh  B365CAHH  B365CAHA  PCAHH  PCAHA  MaxCAHH  \
0     2   A  ...      3.40 -2.25      2.03      1.90   1.99   1.93     2.04   
1     1   D  ...      3.31 -2.25      1.92      2.01   1.92   2.00     1.98   
2     0   D  ...      1.89  0.00      1.92      2.01   1.94   1.97     1.97   
3     2   D  ...      3.58 -2.00      2.07      1.86   2.05   1.86     2.15   
4     1   A  ...      2.26 -0.75      1.92      2.01   1.92   2.00     1.95   

   MaxCAHA  AvgCAHH  AvgCAHA  
0     1.93     1.98     1.91  
1     2.04     1.91     1.97  
2     2.06     1.90     1.99  
3     1.91     2.03     1.85  
4     2.11     1.89        2  

[5 rows x 104 columns]

Wir erstellen eine neue Spalte "FTR_code" in der die Spalte FTR (Full Time Result) als Zahl dargestellt wird, für jedes Spiel.

In [13]:
matches["FTR_code"] = matches["FTR"].astype("category").cat.codes

In [14]:
matches.head()

Date   Time       HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
0 2019-08-16  19:30  Bayern Munich              Hertha     2     2   D     1   
1 2019-08-17  14:30       Dortmund            Augsburg     5     1   H     1   
2 2019-08-17  14:30       Freiburg               Mainz     3     0   H     0   
3 2019-08-17  14:30     Leverkusen           Paderborn     3     2   H     2   
4 2019-08-17  14:30  Werder Bremen  Fortuna Dusseldorf     1     3   A     0   

   HTAG HTR  ...  AHCh  B365CAHH  B365CAHA  PCAHH  PCAHA  MaxCAHH  MaxCAHA  \
0     2   A  ... -2.25      2.03      1.90   1.99   1.93     2.04     1.93   
1     1   D  ... -2.25      1.92      2.01   1.92   2.00     1.98     2.04   
2     0   D  ...  0.00      1.92      2.01   1.94   1.97     1.97     2.06   
3     2   D  ... -2.00      2.07      1.86   2.05   1.86     2.15     1.91   
4     1   A  ... -0.75      1.92      2.01   1.92   2.00     1.95     2.11   

   AvgCAHH  AvgCAHA  FTR_code  
0     1.98     1.91         1  
1     1.91     1.97         2  
2     1.90     1.99         2  
3     2.03     1.85         2  
4     1.89        2         0  

[5 rows x 105 columns]

In [15]:
matches["HTR_codes"] = matches["HTR"].astype("category").cat.codes

In [16]:
matches.head()

Date   Time       HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
0 2019-08-16  19:30  Bayern Munich              Hertha     2     2   D     1   
1 2019-08-17  14:30       Dortmund            Augsburg     5     1   H     1   
2 2019-08-17  14:30       Freiburg               Mainz     3     0   H     0   
3 2019-08-17  14:30     Leverkusen           Paderborn     3     2   H     2   
4 2019-08-17  14:30  Werder Bremen  Fortuna Dusseldorf     1     3   A     0   

   HTAG HTR  ...  B365CAHH  B365CAHA  PCAHH  PCAHA  MaxCAHH  MaxCAHA  AvgCAHH  \
0     2   A  ...      2.03      1.90   1.99   1.93     2.04     1.93     1.98   
1     1   D  ...      1.92      2.01   1.92   2.00     1.98     2.04     1.91   
2     0   D  ...      1.92      2.01   1.94   1.97     1.97     2.06     1.90   
3     2   D  ...      2.07      1.86   2.05   1.86     2.15     1.91     2.03   
4     1   A  ...      1.92      2.01   1.92   2.00     1.95     2.11     1.89   

   AvgCAHA  FTR_code  HTR_codes  
0     1.91         1          0  
1     1.97         2          1  
2     1.99         2          1  
3     1.85         2          1  
4        2         0          0  

[5 rows x 106 columns]

Genauso bekommt jeder Gegner einen Code:

In [17]:
matches["opp_codes"] = matches["AwayTeam"].astype("category").cat.codes

In [18]:
matches.head()

Date   Time       HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
0 2019-08-16  19:30  Bayern Munich              Hertha     2     2   D     1   
1 2019-08-17  14:30       Dortmund            Augsburg     5     1   H     1   
2 2019-08-17  14:30       Freiburg               Mainz     3     0   H     0   
3 2019-08-17  14:30     Leverkusen           Paderborn     3     2   H     2   
4 2019-08-17  14:30  Werder Bremen  Fortuna Dusseldorf     1     3   A     0   

   HTAG HTR  ...  B365CAHA  PCAHH  PCAHA  MaxCAHH  MaxCAHA  AvgCAHH  AvgCAHA  \
0     2   A  ...      1.90   1.99   1.93     2.04     1.93     1.98     1.91   
1     1   D  ...      2.01   1.92   2.00     1.98     2.04     1.91     1.97   
2     0   D  ...      2.01   1.94   1.97     1.97     2.06     1.90     1.99   
3     2   D  ...      1.86   2.05   1.86     2.15     1.91     2.03     1.85   
4     1   A  ...      2.01   1.92   2.00     1.95     2.11     1.89        2   

   FTR_code  HTR_codes  opp_codes  
0         1          0         12  
1         2          1          0  
2         2          1         16  
3         2          1         17  
4         0          0          8  

[5 rows x 107 columns]

In [19]:
matches["team_codes"] = matches["HomeTeam"].astype("category").cat.codes

In [20]:
matches.head()

Date   Time       HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
0 2019-08-16  19:30  Bayern Munich              Hertha     2     2   D     1   
1 2019-08-17  14:30       Dortmund            Augsburg     5     1   H     1   
2 2019-08-17  14:30       Freiburg               Mainz     3     0   H     0   
3 2019-08-17  14:30     Leverkusen           Paderborn     3     2   H     2   
4 2019-08-17  14:30  Werder Bremen  Fortuna Dusseldorf     1     3   A     0   

   HTAG HTR  ...  PCAHH  PCAHA  MaxCAHH  MaxCAHA  AvgCAHH  AvgCAHA  FTR_code  \
0     2   A  ...   1.99   1.93     2.04     1.93     1.98     1.91         1   
1     1   D  ...   1.92   2.00     1.98     2.04     1.91     1.97         2   
2     0   D  ...   1.94   1.97     1.97     2.06     1.90     1.99         2   
3     2   D  ...   2.05   1.86     2.15     1.91     2.03     1.85         2   
4     1   A  ...   1.92   2.00     1.95     2.11     1.89        2         0   

   HTR_codes  opp_codes  team_codes  
0          0         12           1  
1          1          0           5  
2          1         16           9  
3          1         17          14  
4          0          8          22  

[5 rows x 108 columns]

Die Zeit muss auch noch in eine Zahl umgewandelt werden, damit wir damit arbeiten können.

In [21]:
matches["hour"] = matches["Time"].str.replace(":.+", "", regex=True).astype("int")

In [22]:
matches.head()

Date   Time       HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
0 2019-08-16  19:30  Bayern Munich              Hertha     2     2   D     1   
1 2019-08-17  14:30       Dortmund            Augsburg     5     1   H     1   
2 2019-08-17  14:30       Freiburg               Mainz     3     0   H     0   
3 2019-08-17  14:30     Leverkusen           Paderborn     3     2   H     2   
4 2019-08-17  14:30  Werder Bremen  Fortuna Dusseldorf     1     3   A     0   

   HTAG HTR  ...  PCAHA  MaxCAHH  MaxCAHA  AvgCAHH  AvgCAHA  FTR_code  \
0     2   A  ...   1.93     2.04     1.93     1.98     1.91         1   
1     1   D  ...   2.00     1.98     2.04     1.91     1.97         2   
2     0   D  ...   1.97     1.97     2.06     1.90     1.99         2   
3     2   D  ...   1.86     2.15     1.91     2.03     1.85         2   
4     1   A  ...   2.00     1.95     2.11     1.89        2         0   

   HTR_codes  opp_codes  team_codes  hour  
0          0         12           1    19  
1          1          0           5    14  
2          1         16           9    14  
3          1         17          14    14  
4          0          8          22    14  

[5 rows x 109 columns]

Wir fügen noch einen day_code hinzu der die Wochentage als Zahlen kodiert.

In [23]:
matches["day_code"] = matches["Date"].dt.day_of_week

In [24]:
matches.head()

Date   Time       HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
0 2019-08-16  19:30  Bayern Munich              Hertha     2     2   D     1   
1 2019-08-17  14:30       Dortmund            Augsburg     5     1   H     1   
2 2019-08-17  14:30       Freiburg               Mainz     3     0   H     0   
3 2019-08-17  14:30     Leverkusen           Paderborn     3     2   H     2   
4 2019-08-17  14:30  Werder Bremen  Fortuna Dusseldorf     1     3   A     0   

   HTAG HTR  ...  MaxCAHH  MaxCAHA  AvgCAHH  AvgCAHA  FTR_code  HTR_codes  \
0     2   A  ...     2.04     1.93     1.98     1.91         1          0   
1     1   D  ...     1.98     2.04     1.91     1.97         2          1   
2     0   D  ...     1.97     2.06     1.90     1.99         2          1   
3     2   D  ...     2.15     1.91     2.03     1.85         2          1   
4     1   A  ...     1.95     2.11     1.89        2         0          0   

   opp_codes  team_codes  hour  day_code  
0         12           1    19         4  
1          0           5    14         5  
2         16           9    14         5  
3         17          14    14         5  
4          8          22    14         5  

[5 rows x 110 columns]

Wir erstellen eine Spalte "target" mit den Werten 0 und 1. 0 bedeutet, dass die Heimmannschaft verloren hat oder unentschieden, 1 bedeutet, dass die Heimmannschaft gewonnen hat.

In [25]:
matches["target"] = (matches["FTR"] == "H").astype("int")

In [26]:
matches.head()

Date   Time       HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
0 2019-08-16  19:30  Bayern Munich              Hertha     2     2   D     1   
1 2019-08-17  14:30       Dortmund            Augsburg     5     1   H     1   
2 2019-08-17  14:30       Freiburg               Mainz     3     0   H     0   
3 2019-08-17  14:30     Leverkusen           Paderborn     3     2   H     2   
4 2019-08-17  14:30  Werder Bremen  Fortuna Dusseldorf     1     3   A     0   

   HTAG HTR  ...  MaxCAHA  AvgCAHH  AvgCAHA  FTR_code  HTR_codes  opp_codes  \
0     2   A  ...     1.93     1.98     1.91         1          0         12   
1     1   D  ...     2.04     1.91     1.97         2          1          0   
2     0   D  ...     2.06     1.90     1.99         2          1         16   
3     2   D  ...     1.91     2.03     1.85         2          1         17   
4     1   A  ...     2.11     1.89        2         0          0          8   

   team_codes  hour  day_code  target  
0           1    19         4       0  
1           5    14         5       1  
2           9    14         5       1  
3          14    14         5       1  
4          22    14         5       0  

[5 rows x 111 columns]

Wir importieren den RandomForestClassifier aus dem sklearn Paket.

In [27]:
from sklearn.ensemble import RandomForestClassifier

Wir erstellen eine Instanz des Random Forest Klassifikators. n_estimators gibt an, wie viele Bäume wir erstellen wollen. max_depth gibt an, wie tief die Bäume sein dürfen. random_state ist ein Seed, damit wir die Ergebnisse reproduzieren können.

In [28]:
rf = RandomForestClassifier(n_estimators=100, min_samples_split=10, random_state=0)

Wir defnieren unseren Trainings- und Testdatensatz. Unser Trainingsdatensatz sind die Spiele vor dem 27.05.2023. Unsere Testdatensatz sind die Spiele der bisherigen Saison 23/24.

In [29]:
train = matches[matches["Date"] < "27-05-2023"]

In [30]:
test = matches[matches["Date"] > "27-05-2023"]

In [31]:
test.dropna(inplace=True)

/var/folders/lw/pygk9wrs0711t4dprf56xt_m0000gn/T/ipykernel_7708/2920511254.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test.dropna(inplace=True)


In [32]:
train.dropna(inplace=True)

/var/folders/lw/pygk9wrs0711t4dprf56xt_m0000gn/T/ipykernel_7708/167719433.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train.dropna(inplace=True)


Hier definieren wir unsere predictors. Sie sind die Features, die wir für die Vorhersage verwenden wollen. 

In [33]:
predictors = [
    "team_codes", 
    "opp_codes", 
    "hour", 
    "day_code",
    "B365H", "B365D", "B365A", 
    "BWH", "BWD", "BWA", 
    "IWH", "IWD", "IWA", 
    "PSH", "PSD", "PSA", 
    "WHH", "WHD", "WHA", 
    "VCH", "VCD", "VCA", 
    "MaxH", "MaxD", "MaxA"
]


Mit rf.fit trainieren wir das Modell mit den Trainingsdaten.

In [34]:
rf.fit(train[predictors], train["target"])

RandomForestClassifier(min_samples_split=10, random_state=0)

preds ist hier die Variable, die die Vorhersagen des Modells enthält: 

In [35]:
preds = rf.predict(test[predictors])

In [36]:
from sklearn.metrics import accuracy_score

Die Variable acc ist die Genauigkeit des Modells. Sie gibt an, wie oft das Modell richtig vorhersagt, ob die Heimmannschaft gewinnt oder nicht:

In [37]:
acc = accuracy_score(test["target"], preds)

In [38]:
acc

0.6630434782608695

In [39]:
combined = pd.DataFrame(dict(actual=test["target"], prediction=preds))

In [40]:
pd.crosstab(index=combined["actual"], columns=combined["prediction"])

prediction   0   1
actual            
0           38  12
1           19  23

In [41]:
from sklearn.metrics import precision_score

In [42]:
precision_score(test["target"], preds)

0.6571428571428571

In [43]:
grouped_matches = matches.groupby("HomeTeam")

In [44]:
group = grouped_matches.get_group("Bayern Munich")

In [45]:
group

Date   Time       HomeTeam      AwayTeam  FTHG  FTAG FTR  HTHG  \
0    2019-08-16  19:30  Bayern Munich        Hertha     2     2   D     1   
19   2019-08-31  14:30  Bayern Munich         Mainz     6     1   H     2   
37   2019-09-21  14:30  Bayern Munich       FC Koln     4     0   H     1   
55   2019-10-05  14:30  Bayern Munich    Hoffenheim     1     2   A     0   
73   2019-10-26  14:30  Bayern Munich  Union Berlin     2     1   H     1   
...         ...    ...            ...           ...   ...   ...  ..   ...   
1251 2023-09-15  19:30  Bayern Munich    Leverkusen     2     2   D     1   
1262 2023-09-23  14:30  Bayern Munich        Bochum     7     0   H     4   
1285 2023-10-08  16:30  Bayern Munich      Freiburg     3     0   H     2   
1298 2023-10-28  14:30  Bayern Munich     Darmstadt     8     0   H     0   
1316 2023-11-11  14:30  Bayern Munich    Heidenheim     4     2   H     2   

      HTAG HTR  ...  MaxCAHA  AvgCAHH  AvgCAHA  FTR_code  HTR_codes  \
0        2   A  ...     1.93     1.98     1.91         1          0   
19       1   H  ...     1.90     2.02     1.86         2          2   
37       0   H  ...     2.08     1.86     2.02         2          2   
55       0   D  ...     2.03     1.93     1.95         0          1   
73       0   H  ...     1.99     1.95     1.93         2          2   
...    ...  ..  ...      ...      ...      ...       ...        ...   
1251     1   D  ...     2.04     1.92     1.94         1          1   
1262     0   H  ...     2.02     1.94      1.9         2          2   
1285     0   H  ...     2.19     1.86        2         2          2   
1298     0   D  ...     2.05     1.89     1.96         2          1   
1316     0   H  ...     2.14     1.92     1.92         2          2   

      opp_codes  team_codes  hour  day_code  target  
0            12           1    19         4       0  
19           16           1    14         5       1  
37            7           1    14         5       1  
55           13           1    14         5       0  
73           21           1    14         5       1  
...         ...         ...   ...       ...     ...  
1251         14           1    19         4       0  
1262          3           1    14         5       1  
1285          9           1    16         6       1  
1298          4           1    14         5       1  
1316         11           1    14         5       1  

[74 rows x 111 columns]

In [46]:
def rolling_average(group, cols, new_cols):
    group = group.sort_values("Date")
    rolling_stats = group[cols].rolling(5, closed="left").mean()
    group[new_cols] = rolling_stats
    group = group.dropna(subset=new_cols)
    return group

In [47]:
matches.columns

Index(['Date', 'Time', 'HomeTeam', 'AwayTeam', 'FTHG', 'FTAG', 'FTR', 'HTHG',
       'HTAG', 'HTR',
       ...
       'MaxCAHA', 'AvgCAHH', 'AvgCAHA', 'FTR_code', 'HTR_codes', 'opp_codes',
       'team_codes', 'hour', 'day_code', 'target'],
      dtype='object', length=111)

In [48]:
cols = ["FTHG", "FTAG", "HTHG", "HTAG", "HS", "AS", "HST", "AST", "HC", "AC", "HF", "AF", "HY", "AY", "HR", "AR"]
new_cols = [f"{c}_rolling" for c in cols]

In [49]:
new_cols

['FTHG_rolling',
 'FTAG_rolling',
 'HTHG_rolling',
 'HTAG_rolling',
 'HS_rolling',
 'AS_rolling',
 'HST_rolling',
 'AST_rolling',
 'HC_rolling',
 'AC_rolling',
 'HF_rolling',
 'AF_rolling',
 'HY_rolling',
 'AY_rolling',
 'HR_rolling',
 'AR_rolling']

In [50]:
rolling_average(group, cols, new_cols)

Date   Time       HomeTeam       AwayTeam  FTHG  FTAG FTR  HTHG  \
95   2019-11-09  17:30  Bayern Munich       Dortmund     4     0   H     1   
113  2019-11-30  17:30  Bayern Munich     Leverkusen     1     2   A     1   
127  2019-12-14  14:30  Bayern Munich  Werder Bremen     6     1   H     2   
145  2019-12-21  14:30  Bayern Munich      Wolfsburg     2     0   H     0   
168  2020-01-25  17:30  Bayern Munich     Schalke 04     5     0   H     2   
...         ...    ...            ...            ...   ...   ...  ..   ...   
1251 2023-09-15  19:30  Bayern Munich     Leverkusen     2     2   D     1   
1262 2023-09-23  14:30  Bayern Munich         Bochum     7     0   H     4   
1285 2023-10-08  16:30  Bayern Munich       Freiburg     3     0   H     2   
1298 2023-10-28  14:30  Bayern Munich      Darmstadt     8     0   H     0   
1316 2023-11-11  14:30  Bayern Munich     Heidenheim     4     2   H     2   

      HTAG HTR  ...  HST_rolling  AST_rolling  HC_rolling  AC_rolling  \
95       0   H  ...          7.2          3.6        10.2         2.0   
113      2   A  ...          6.8          3.0         8.2         2.2   
127      1   H  ...          7.0          3.6         7.8         2.4   
145      0   D  ...          8.8          3.6         7.0         2.0   
168      0   H  ...          9.6          3.2         7.6         2.6   
...    ...  ..  ...          ...          ...         ...         ...   
1251     1   D  ...          8.0          2.8        10.4         2.8   
1262     0   H  ...          8.4          3.2         9.2         2.8   
1285     0   H  ...          9.8          3.2         7.2         2.8   
1298     0   D  ...          8.0          3.2         7.8         2.4   
1316     0   H  ...         10.4          2.4         7.8         2.2   

      HF_rolling  AF_rolling  HY_rolling  AY_rolling  HR_rolling  AR_rolling  
95          10.8        15.2         2.4         1.4         0.0         0.2  
113         11.8        14.2         2.2         1.0         0.0         0.2  
127         10.4        15.0         1.8         1.0         0.0         0.4  
145         10.0        13.4         1.4         1.4         0.0         0.2  
168          9.8        12.4         1.4         1.8         0.0         0.2  
...          ...         ...         ...         ...         ...         ...  
1251         8.6         9.4         1.2         1.4         0.0         0.0  
1262         8.4         8.6         0.8         1.8         0.0         0.0  
1285         9.0         9.0         0.4         2.0         0.0         0.0  
1298         8.4         8.6         0.6         1.4         0.0         0.0  
1316         9.8         9.0         0.6         1.6         0.2         0.4  

[69 rows x 127 columns]

In [51]:
matches_rolling = matches.groupby("HomeTeam").apply(lambda x: rolling_average(x, cols, new_cols))

In [52]:
matches_rolling

Date   Time   HomeTeam            AwayTeam  FTHG  FTAG  \
HomeTeam                                                                      
Augsburg  106  2019-11-24  14:30   Augsburg              Hertha     4     0   
          118  2019-12-07  14:30   Augsburg               Mainz     2     1   
          136  2019-12-17  19:30   Augsburg  Fortuna Dusseldorf     3     0   
          154  2020-01-18  14:30   Augsburg            Dortmund     3     5   
          172  2020-02-01  14:30   Augsburg       Werder Bremen     2     1   
...                   ...    ...        ...                 ...   ...   ...   
Wolfsburg 1256 2023-09-16  14:30  Wolfsburg        Union Berlin     2     1   
          1274 2023-09-30  14:30  Wolfsburg       Ein Frankfurt     2     0   
          1292 2023-10-21  14:30  Wolfsburg          Leverkusen     1     2   
          1312 2023-11-05  14:30  Wolfsburg       Werder Bremen     2     2   
          1328 2023-11-25  14:30  Wolfsburg          RB Leipzig     2     1   

               FTR  HTHG  HTAG HTR  ...  HST_rolling  AST_rolling  HC_rolling  \
HomeTeam                            ...                                         
Augsburg  106    H     2     0   H  ...          4.4          4.2         2.0   
          118    H     1     1   D  ...          4.6          4.2         2.0   
          136    H     1     0   H  ...          4.4          4.2         3.2   
          154    A     1     0   H  ...          5.6          3.6         3.8   
          172    H     0     1   A  ...          5.2          3.4         3.6   
...             ..   ...   ...  ..  ...          ...          ...         ...   
Wolfsburg 1256   H     2     1   H  ...          5.4          2.8         3.2   
          1274   H     1     0   H  ...          5.4          3.0         3.4   
          1292   A     1     1   D  ...          6.0          3.2         4.2   
          1312   D     1     1   D  ...          6.8          4.0         4.4   
          1328   H     1     0   H  ...          6.0          3.8         4.0   

                AC_rolling  HF_rolling  AF_rolling  HY_rolling  AY_rolling  \
HomeTeam                                                                     
Augsburg  106          6.2        12.4        10.4         3.0         2.0   
          118          6.0        10.8        10.2         2.4         2.2   
          136          4.2        11.8        11.4         2.4         2.4   
          154          4.0        10.6        12.6         2.2         3.2   
          172          3.8         9.4        12.4         1.6         2.8   
...                    ...         ...         ...         ...         ...   
Wolfsburg 1256         5.6        10.4        13.2         1.4         3.0   
          1274         5.8         9.6        11.4         1.6         2.0   
          1292         5.6        10.6        13.0         1.6         2.6   
          1312         4.8        10.2        13.4         1.8         3.2   
          1328         4.6        11.2        13.2         2.4         2.4   

                HR_rolling  AR_rolling  
HomeTeam                                
Augsburg  106          0.0         0.2  
          118          0.0         0.2  
          136          0.0         0.2  
          154          0.0         0.2  
          172          0.0         0.2  
...                    ...         ...  
Wolfsburg 1256         0.0         0.0  
          1274         0.0         0.0  
          1292         0.0         0.2  
          1312         0.0         0.2  
          1328         0.2         0.2  

[1212 rows x 127 columns]

In [53]:
matches_rolling = matches_rolling.droplevel("HomeTeam")

In [54]:
matches_rolling

Date   Time   HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
106  2019-11-24  14:30   Augsburg              Hertha     4     0   H     2   
118  2019-12-07  14:30   Augsburg               Mainz     2     1   H     1   
136  2019-12-17  19:30   Augsburg  Fortuna Dusseldorf     3     0   H     1   
154  2020-01-18  14:30   Augsburg            Dortmund     3     5   A     1   
172  2020-02-01  14:30   Augsburg       Werder Bremen     2     1   H     0   
...         ...    ...        ...                 ...   ...   ...  ..   ...   
1256 2023-09-16  14:30  Wolfsburg        Union Berlin     2     1   H     2   
1274 2023-09-30  14:30  Wolfsburg       Ein Frankfurt     2     0   H     1   
1292 2023-10-21  14:30  Wolfsburg          Leverkusen     1     2   A     1   
1312 2023-11-05  14:30  Wolfsburg       Werder Bremen     2     2   D     1   
1328 2023-11-25  14:30  Wolfsburg          RB Leipzig     2     1   H     1   

      HTAG HTR  ...  HST_rolling  AST_rolling  HC_rolling  AC_rolling  \
106      0   H  ...          4.4          4.2         2.0         6.2   
118      1   D  ...          4.6          4.2         2.0         6.0   
136      0   H  ...          4.4          4.2         3.2         4.2   
154      0   H  ...          5.6          3.6         3.8         4.0   
172      1   A  ...          5.2          3.4         3.6         3.8   
...    ...  ..  ...          ...          ...         ...         ...   
1256     1   H  ...          5.4          2.8         3.2         5.6   
1274     0   H  ...          5.4          3.0         3.4         5.8   
1292     1   D  ...          6.0          3.2         4.2         5.6   
1312     1   D  ...          6.8          4.0         4.4         4.8   
1328     0   H  ...          6.0          3.8         4.0         4.6   

      HF_rolling  AF_rolling  HY_rolling  AY_rolling  HR_rolling  AR_rolling  
106         12.4        10.4         3.0         2.0         0.0         0.2  
118         10.8        10.2         2.4         2.2         0.0         0.2  
136         11.8        11.4         2.4         2.4         0.0         0.2  
154         10.6        12.6         2.2         3.2         0.0         0.2  
172          9.4        12.4         1.6         2.8         0.0         0.2  
...          ...         ...         ...         ...         ...         ...  
1256        10.4        13.2         1.4         3.0         0.0         0.0  
1274         9.6        11.4         1.6         2.0         0.0         0.0  
1292        10.6        13.0         1.6         2.6         0.0         0.2  
1312        10.2        13.4         1.8         3.2         0.0         0.2  
1328        11.2        13.2         2.4         2.4         0.2         0.2  

[1212 rows x 127 columns]

In [55]:
matches_rolling.index = range(matches_rolling.shape[0])

In [56]:
matches_rolling

Date   Time   HomeTeam            AwayTeam  FTHG  FTAG FTR  HTHG  \
0    2019-11-24  14:30   Augsburg              Hertha     4     0   H     2   
1    2019-12-07  14:30   Augsburg               Mainz     2     1   H     1   
2    2019-12-17  19:30   Augsburg  Fortuna Dusseldorf     3     0   H     1   
3    2020-01-18  14:30   Augsburg            Dortmund     3     5   A     1   
4    2020-02-01  14:30   Augsburg       Werder Bremen     2     1   H     0   
...         ...    ...        ...                 ...   ...   ...  ..   ...   
1207 2023-09-16  14:30  Wolfsburg        Union Berlin     2     1   H     2   
1208 2023-09-30  14:30  Wolfsburg       Ein Frankfurt     2     0   H     1   
1209 2023-10-21  14:30  Wolfsburg          Leverkusen     1     2   A     1   
1210 2023-11-05  14:30  Wolfsburg       Werder Bremen     2     2   D     1   
1211 2023-11-25  14:30  Wolfsburg          RB Leipzig     2     1   H     1   

      HTAG HTR  ...  HST_rolling  AST_rolling  HC_rolling  AC_rolling  \
0        0   H  ...          4.4          4.2         2.0         6.2   
1        1   D  ...          4.6          4.2         2.0         6.0   
2        0   H  ...          4.4          4.2         3.2         4.2   
3        0   H  ...          5.6          3.6         3.8         4.0   
4        1   A  ...          5.2          3.4         3.6         3.8   
...    ...  ..  ...          ...          ...         ...         ...   
1207     1   H  ...          5.4          2.8         3.2         5.6   
1208     0   H  ...          5.4          3.0         3.4         5.8   
1209     1   D  ...          6.0          3.2         4.2         5.6   
1210     1   D  ...          6.8          4.0         4.4         4.8   
1211     0   H  ...          6.0          3.8         4.0         4.6   

      HF_rolling  AF_rolling  HY_rolling  AY_rolling  HR_rolling  AR_rolling  
0           12.4        10.4         3.0         2.0         0.0         0.2  
1           10.8        10.2         2.4         2.2         0.0         0.2  
2           11.8        11.4         2.4         2.4         0.0         0.2  
3           10.6        12.6         2.2         3.2         0.0         0.2  
4            9.4        12.4         1.6         2.8         0.0         0.2  
...          ...         ...         ...         ...         ...         ...  
1207        10.4        13.2         1.4         3.0         0.0         0.0  
1208         9.6        11.4         1.6         2.0         0.0         0.0  
1209        10.6        13.0         1.6         2.6         0.0         0.2  
1210        10.2        13.4         1.8         3.2         0.0         0.2  
1211        11.2        13.2         2.4         2.4         0.2         0.2  

[1212 rows x 127 columns]

In [57]:
def make_prediction(data, predictors):
    train = data[data["Date"] < "27-05-2023"]
    test = data[data["Date"] > "27-05-2023"]
    rf.fit(train[predictors], train["target"])
    preds = rf.predict(test[predictors])
    combined = pd.DataFrame(dict(actual=test["target"], prediction=preds))
    precision = precision_score(test["target"], preds)
    return combined, precision

In [58]:
matches_rolling.dropna(inplace=True)

In [59]:
combined, precision = make_prediction(matches_rolling, predictors + new_cols)

In [60]:
precision

0.7777777777777778

In [61]:
combined

actual  prediction
63         0           0
64         0           0
65         1           0
66         0           0
67         1           0
...      ...         ...
1207       1           1
1208       1           0
1209       0           0
1210       0           0
1211       1           0

[84 rows x 2 columns]

In [62]:
import pickle

In [63]:
with open ("../models/model.pkl", "wb") as f:
    pickle.dump(rf, f)